aim/db/hashtree_db_listener.py

```diff
@@ -202,6 +202,33 @@ def _preprocess_logs(self, logs):
             if not aim_res:
                 LOG.warn('Aim resource for event %s not found' % log)
                 continue
+            # REVISIT: We currently only query the DB for
+            # SecurityGroupRule resources, but should treat all
+            # resource types uniformly, and therefore should do this
+            # for all resource types. This will also allow elimination
+            # of the epoch bumping when modifying list attributes of
+            # other resource types. But we probably will want to
+            # optimize this quering to avoid a roundtrip to the DB
+            # server for each action log item, by making one find()
+            # call for each resource type, filtering with the
+            # identities from all the action log items being
+            # processed.
+            if isinstance(aim_res, resource.SecurityGroupRule):
+                db_aim_res = self.aim_manager.get(ctx, aim_res)
+                if db_aim_res:
+                    if action == aim_tree.ActionLog.DELETE:
+                        LOG.warn("AIM resource %s exists in DB for delete "
+                                 "action" % db_aim_res)
+                    else:
+                        # Use current resource from DB so that list
+                        # attributes do no need to be protected from
+                        # concurrent updates by bumping the resource's
+                        # epoch.
+                        aim_res = db_aim_res
+                else:
+                    if action != aim_tree.ActionLog.DELETE:
+                        LOG.warn("AIM resource %s does not exist in DB for "
```

Comment:
The warning log will only make sense if its for the UPDATE action I think. Because for CREATE action, it makes sense that the record is not in the DB yet.